File: scala-package/core/src/main/scala/ml/dmlc/mxnet/contrib/AutoGrad.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ml.dmlc.mxnet.contrib

import ml.dmlc.mxnet.Base._
import ml.dmlc.mxnet._
import scala.collection.mutable.ArrayBuffer

object AutoGrad {

  /**
   * Turn on or turn off operator recording.
   *
   * @param recording Boolean
   */
  def setRecording(recording: Boolean): Unit = {
    val flag = if (recording) 1 else 0
    checkCall(_LIB.mxAutogradSetRecording(flag))
  }

  /**
  * Mark NDArrays as variables to compute gradient for autograd.
  *
  * @param variables array of NDArray
  */
  def markVariables(variables: Array[NDArray]): Unit = {
    val variableHandles = variables.map(_.handle)
    checkCall(_LIB.mxAutogradMarkVariables(variableHandles))
  }

  /**
  * Compute the gradients of outputs w.r.t variables.
  *
  * @param outputs array of NDArray
  * @return gradients array of NDArray
  */
  def computeGradient(outputs: Array[NDArray]): Array[NDArray] = {
    val outputHandles = outputs.map(_.handle)

    val gradHandles = ArrayBuffer.empty[NDArrayHandle]
    checkCall(_LIB.mxAutogradComputeGradient(outputHandles, gradHandles))
    gradHandles.map(new NDArray(_)).toArray
  }

  /**
  * Return function that computes both gradient of arguments and loss value.
  *
  * @param func a Scala function
  *                 The forward (loss) function.
  * @return gradAndLossFunc a Scala function
  *                A function that would compute both the gradient of arguments and loss value.
  */
  def gradAndLoss(func: Array[NDArray] => Array[NDArray]):
    Array[NDArray] => (Array[NDArray], Array[NDArray]) = {
    // Wrapped function
    def wrapped(args: Array[NDArray]): (Array[NDArray], Array[NDArray]) = {
      markVariables(args)
      setRecording(true)
      val outputs = func(args)
      setRecording(false)
      val gradVals = computeGradient(outputs)
      (gradVals, outputs)
    }
    wrapped
  }

  /**
  * Return function that computes gradient of arguments.
  *
  * @param func a Scala function
  *                 The forward (loss) function.
  * @return gradFunc a Scala function
  *                A function that would compute the gradient of arguments.
  */
  def grad(func: Array[NDArray] => Array[NDArray]): Array[NDArray] => Array[NDArray] = {
    def wrapped(args: Array[NDArray]): Array[NDArray] = {
      gradAndLoss(func)(args)._1
    }
    wrapped
  }
}

File: csharp-package/src/MxNet/Autograd.cs
/*****************************************************************************
   Copyright 2018 The MxNet.Sharp Authors. All Rights Reserved.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
******************************************************************************/
using System;
using System.Linq;
using MxNet.Interop;
using MxNet.Numpy;
using MxNet.Sym.Numpy;

namespace MxNet
{
    public class Autograd
    {
        public static bool SetRecording(bool is_recording)
        {
            var prev = 0;
            NativeMethods.MXAutogradSetIsRecording(Convert.ToInt32(is_recording), ref prev);

            return Convert.ToBoolean(prev);
        }

        public static bool SetTraining(bool train_mode)
        {
            var prev = 0;
            NativeMethods.MXAutogradSetIsTraining(Convert.ToInt32(train_mode), ref prev);

            return Convert.ToBoolean(prev);
        }

        public static bool IsRecording()
        {
            var curr = 0;
            NativeMethods.MXAutogradIsRecording(ref curr);

            return Convert.ToBoolean(curr);
        }

        public static bool IsTraining()
        {
            var curr = 0;
            NativeMethods.MXAutogradIsTraining(ref curr);

            return Convert.ToBoolean(curr);
        }

        public static _RecordingStateScope Record(bool train_mode = true)
        {
            return new _RecordingStateScope(true, train_mode);
        }

        public static _RecordingStateScope Pause(bool train_mode = true)
        {
            return new _RecordingStateScope(false, train_mode);
        }

        public static _RecordingStateScope TrainMode()
        {
            return new _RecordingStateScope(null, true);
        }

        public static _RecordingStateScope PredictMode()
        {
            return new _RecordingStateScope(null, false);
        }

        public static void MarkVariables(NDArrayList variables, NDArrayList gradients,
            OpGradReq grad_reqs = OpGradReq.Write)
        {
            var gradReqs = new int[variables.Length];
            for (var i = 0; i < gradReqs.Length; i++) gradReqs[i] = (int) OpGradReq.Write;

            NativeMethods.MXAutogradMarkVariables(variables.Length, MxUtil.GetNDArrayHandles(variables), gradReqs,
                MxUtil.GetNDArrayHandles(gradients));
        }

        private static (IntPtr[], IntPtr[]) ParseHead(NDArrayList heads, NDArrayList head_grads)
        {
            IntPtr[] headHandles = null;
            IntPtr[] headGradHandles = null;

            headHandles = MxUtil.GetNDArrayHandles(heads);

            if (head_grads == null)
            {
                headGradHandles = new IntPtr[heads.Length];
                for (var i = 0; i < headHandles.Length; i++) headGradHandles[i] = IntPtr.Zero;
            }
            else
            {
                if (heads.Length != head_grads.Length)
                    throw new ArgumentException("heads and head_grads must be lists of the same length");

                headGradHandles = MxUtil.GetNDArrayHandles(head_grads);
            }

            return (headHandles, headGradHandles);
        }

        public static void Backward(NDArrayList heads, NDArrayList head_grads = null, bool retain_graph = false,
            bool train_mode = true)
        {
            var (head_handles, head_grads_handles) = ParseHead(heads, head_grads);

            NativeMethods.MXAutogradBackwardEx(head_handles.Length, head_handles, head_grads_handles, 0,
                new IntPtr[] { }, Convert.ToInt32(retain_graph),
                0, Convert.ToInt32(train_mode), out var grad_handles, out var grad_count);
        }

        public static NDArrayList Grad(NDArrayList heads, NDArrayList variables, NDArrayList head_grads = null,
            bool retain_graph = false, bool create_graph = true, bool train_mode = true)
        {
            var (head_handles, head_grads_handles) = ParseHead(heads, head_grads);

            //var grad_handles = new IntPtr[head_handles.Length];
            //var grad_stypes = new int[head_handles.Length];

            NativeMethods.MXAutogradBackwardEx(head_handles.Length, head_handles, head_grads_handles, variables.Length,
                MxUtil.GetNDArrayHandles(variables), Convert.ToInt32(retain_graph),
                Convert.ToInt32(create_graph), Convert.ToInt32(train_mode), out var grad_handles, out var grad_stypes);

            var result = new NDArrayList();
            foreach (var item in grad_handles) result.Add(new NDArray(item));

            return result.ToArray();
        }

        internal static _Symbol GetSymbol(ndarray x)
        {
            var hdl = IntPtr.Zero;
            NativeMethods.MXAutogradGetSymbol(x.GetHandle(), hdl);
            return new _Symbol(hdl);
        }

        public class _RecordingStateScope : MxDisposable
        {
            private bool? _enter_is_record;
            private bool? _enter_train_mode;
            private bool? _prev_is_record;
            private bool? _prev_train_mode;

            public _RecordingStateScope(bool? is_record, bool train_mode)
            {
                _enter_is_record = is_record;
                _enter_train_mode = train_mode;
                _prev_is_record = null;
                _prev_train_mode = null;
                With();
            }

            public override MxDisposable With()
            {
                if (_enter_is_record.HasValue)
                    _prev_is_record = SetRecording(_enter_is_record.Value);

                if (_enter_train_mode.HasValue)
                    _prev_train_mode = SetTraining(_enter_train_mode.Value);
                return this;
            }

            public override void Exit()
            {
                if (_enter_is_record.HasValue && _prev_is_record != _enter_is_record)
                    SetRecording(_prev_is_record.Value);

                if (_enter_train_mode.HasValue && _prev_train_mode != _enter_train_mode)
                    SetTraining(_prev_train_mode.Value);
            }
        }
    }
}